[crypto] introduce PSA Crypto API backend

LuDuda opened the pull request:
This PR introduces initial support for the PSA Crypto API throughout OpenThread (#10253), in preparation for the deprecation of the existing mbedTLS API in mbedTLS 4.0 (see [long-term plan](https://mbed-tls.readthedocs.io/en/latest/project/long-term-plans/#making-psa-crypto-the-only-crypto-api))​. 

The PR is a bit large; however, it includes the minimal set of changes necessary to pass unit tests, expect tests, and certification tests. Commits have been split into smaller logical units to make review easier.

There are a few items not included in this PR (to keep it reviewable), which I plan to contribute after this one is merged like: Support for RCP, POSIX or ot-br-posix integration.

I have left the mbedTLS backend as the default for now, however per the mbedTLS team’s recommendation, we should switch to PSA Crypto API as the default in the near future.

**This PR addresses the following items:**

1. Extending OpenThread platform API to support HKDF-SHA256  in PSA API (needed by TREL)
2. Introducing default implementation of the OpenThread Crypto Platform API using the pure PSA Crypto API
3. Refactoring the `mbedtls-config.h` file.
4. Adjusting unit tests for AES, ECDSA, HMAC-SHA256, and HKDF-SHA256 to work with PSA Crypto
5. Introducing platform support for PSA Crypto API, which includes:
- RNG support via PSA
- Native ITS support for persistent keys
6. Enabling CI checks for the simulation platform with PSA Crypto, covering:
 - Unit tests
 - Thread certification & system tests
 - Expect tests
7. Enhancements to the overall codebase necessary to pass certain tests

**Note for reviewers:**
Please pay special attention to the Native ITS implementation. I initially tried to reuse the existing settings/flash APIs but found it challenging to satisfy both the simulation and POSIX platforms and CI. Any feedback on this design choice would be greatly appreciated. The Native ITS file implementation might need some enhancements later.

FYI: @gilles-peskine-arm @hemanth-silabs

LuDuda committed:
[crypto] PSA API: align AES unit tests

This commit allows running AES unit tests using PSA Crypto API.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

LuDuda committed:
[crypto] PSA API: align HMAC-SHA256 unit tests

This commit allows running HMAC-SHA256 unit tests using PSA Crypto API.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

LuDuda committed:
[crypto] PSA API: align HKDF-SHA256 unit tests

This commit allows running HKDF-SHA256 unit tests using PSA Crypto API.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

LuDuda committed:
[crypto] PSA API: align ECDSA unit tests

This commit allows running ECDSA unit tests using PSA Crypto API.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

LuDuda committed:
[crypto] PSA API: introduce example native ITS implementations

This commit implements two example native ITS for linux-based platform.

LuDuda committed:
[crypto] PSA API: extend Crypto platform API to support HKDF-SHA256

This commit adds PSA API support for HKDF-SHA256 for TREL in OpenThread
platform API.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

LuDuda committed:
[crypto] PSA API: introduce default OpenThread crypto platform port

This commit adds a default implementation of the OpenThread crypto
platform API for the PSA Crypto API.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

LuDuda committed:
[crypto] PSA API: introduce support for simulation and test platform

This commit adds support for Crypto PSA library for simulation and test
platform and introduces initial CI check.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

LuDuda committed:
[crypto] PSA API: set platform key references for PSA automatically

This commit configures the
OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE based on the chosen
crypto library.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

LuDuda committed:
[crypto] PSA API: enable native ITS in build system

This commit enables PSA native ITS implementation in the build system
for simulation and test platforms.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

LuDuda committed:
[crypto] PSA API: add basic unit, cert and expect testing in CI

This commit introduces initial CI testing of PSA Crypto variant.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

Damian-Nordic reviewed (approved):
I already reviewed these changes internally.

abtink commented on src/core/thread/key_manager.cpp:
```diff
@@ -313,7 +313,17 @@ void KeyManager::ComputeTrelKey(uint32_t aKeySequence, Mac::Key &aKey) const
+                                               Crypto::Storage::kKeyAlgorithmHkdfSha256, Crypto::Storage::kUsageDerive,
+                                               Crypto::Storage::kTypeVolatile, networkKey.m8, NetworkKey::kSize));
+
+    cryptoKey.SetAsKeyRef(keyRef);
```
I've noticed this pattern in several places where we seem to export a key and then save it again as a new `keyref`.

Why do we need to create a new `keyref` in these situations? Can the existing `mNetworkKeyRef` be configured to allow additional crypto algorithms (like `kKeyAlgorithmHkdfSha256`)?

If we have to export/read the key and then import/write it back as a new `keyref`, doesn't that somewhat defeat the purpose of secure storage?

Is there a way to specify that `mNetworkKeyRef` can be used with a particular set of crypto algorithms?

If this is not possible, alternatively, could the crypto platform implementation handle the `keyref` copying? For example, `otPlatCryptoHkdfExtract()` could check if `aInputKey` is usable for the intended operation, and if not, perform the `keyref` copying itself.
I feel this might be a better approach than having the OT core manage this.

LuDuda replied:
Yes, introduction of new crypto platform API to make a copy is correct way - I had todo comments before i raised this PR about it, but decided to leave it, as Network Key today is anyway exportable. Will fix it in this PR.

LuDuda replied:
After playing a bit, It does not seem possible to perform both HMAC-SHA256 (Sign Hash) and HKDF-SHA256 (Derive) on a single key reference, nor to use PSA Key Copy on different key types - unless I'm missing something. Any thoughts? @gilles-peskine-arm @hemanth-silabs

Therefore, the only viable options I see are:
1. Export and re-import the key in the OpenThread code (as shown in this PR).
2. Export and re-import the key in the platform code (in HKDF implementation)
3. Implement `otPlatCryptoKeyCopy` as i intended.. but use export&import pattern if key types do not match.
4. When initially importing (provisioning) the Network Key as a persistent key, also create a separate persistent key for HKDF purpose.

If there aren’t any better alternatives, I can go ahead and implement Abtin’s suggested approach or in fact what i intended with point 3 (but not using `psa_copy_key` function).

gilles-peskine-arm replied:
The PSA API has a policy on keys that restricts each key to one algorithm. (There are a few exceptions but they don't apply here.) So if you need to use the same key material for both authentication with HMAC and key derivation with HKDF, you have to create two different key identifiers (meaning that there are two copies of the key material under the hood).

A protocol that uses the same key for HMAC and HKDF is dangerous (you have to be very careful not to disclose derived keys when calculating an authentication tag for a message). But I guess you don't have a choice...

Unless storage space is the most important resource, I would suggest (4), so you don't have to worry about keys being exportable. But export-import at the point of use (1, 2 or 3 — the differences are a matter of OpenThread design which I don't know about) also works.

`psa_copy_key` won't help because it respects policies. This function is mostly intended to transfer keys between applications in multi-application environments, it's rarely useful in single-application environments.

LuDuda replied:
Updated the PR, with option 2 implemented as suggested by Abtin.

abtink commented on src/core/mac/link_raw.cpp:
```diff
@@ -58,6 +58,8 @@ LinkRaw::LinkRaw(Instance &aInstance)
     , mSubMac(aInstance.Get<SubMac>())
 #endif
 {
+    otPlatCryptoInit();
```
Do we need this call to `otPlatCryptoInit()` here?

The `otPlatCryptoInit()` is already called from the `KeyManager` constructor.

I assume this was added here for the RCP build. However, `LinkRaw` can also be included in an FTD/MTD build, which would result in `otPlatCryptoInit()` being called twice.

A better approach can be to move the `otPlatCryptoInit()` call to the `Instance` constructor, ensuring it is called once for all build types (FTD/MTD or RCP)?

abtink commented on tests/unit/test_dns_client.cpp:
```diff
@@ -1385,6 +1385,11 @@ void TestDnsClient(void)
 
+#if (OPENTHREAD_CONFIG_CRYPTO_LIB == OPENTHREAD_CONFIG_CRYPTO_LIB_PSA) && OPENTHREAD_CONFIG_HEAP_EXTERNAL_ENABLE
+    // On a first attempt, SRP Client generates the SRP Key which adds additional heap allocation.
+    heapAllocations += 1;
```
I noticed this change and its associated comment are added to several tests.

I assume this is necessary because MbedTLS is allocating some heap memory (under `CRYPTO_LIB_PSA`) and not freeing it by the end of the test. It also seems that MbedTLS is configured to use `otPlatCAlloc()`.

While I understand this change and see that it allows the unit tests to pass, I think this model could be rather fragile.

The purpose of `otPlatCAlloc` and `otPlatFree` is to serve as platform-provided heap allocation APIs for OpenThread core modules. That's why, in these unit tests, we track their usage to ensure there are no memory leaks from the core.

Could we consider defining a new set of `otPlatCrypto` APIs for heap allocation, specifically for the crypto module, which could then be used by the crypto library (e.g., MbedTLS)?

I think this approach would allow us to decouple the heap APIs used by different components and potentially avoid the need to handle these memory tracking exceptions in tests.
What do you think?

hemanth-silabs replied:
I agree, can we instead implement copy key @LuDuda? This way, PSA core will manage creating copy of the key wit new attributes.

LuDuda replied:
@abtink right, i also don't particularly like this code either and was sure about getting a comment :) indeed MbedTLS heap allocation was necessary by design during allocation of volatile key. I like the idea of decoupling otPlatCalloc with otPlatCryptoCalloc - will take a look and propose some changes.

LuDuda replied:
> I agree, can we instead implement copy key @LuDuda? This way, PSA core will manage creating copy of the key wit new attributes.

I think you have referred to the previous comment. Yes, i will add a new copy API - this is correct way.

gilles-peskine-arm replied:
I'm not familiar with OpenThread's architecture, but I can give some information on how Mbed TLS manages memory. I hope this helps.

You can configure which `calloc` and `free` functions Mbed TLS uses by defining [`MBEDTLS_PLATFORM_MEMORY`](https://mbed-tls.readthedocs.io/projects/api/en/v3.6.3/api/file/mbedtls__config_8h/#c.MBEDTLS_PLATFORM_MEMORY) and the relevant additional macros. (Note that Mbed TLS only wants `calloc` and `free`, not `malloc` nor `realloc`.)

`psa_crypto_init()` allocates heap memory (in most configurations). The function [`mbedtls_psa_crypto_free()`](https://mbed-tls.readthedocs.io/projects/api/en/v3.6.3/api/file/crypto__extra_8h/#_CPPv423mbedtls_psa_crypto_freev) frees this memory. Beyond that, creating a key (e.g. with `psa_import_key()`, `psa_generate_key()`, etc.) may allocate memory (depending on the configuration and the key type), and `psa_destroy_key()` frees this memory. (When `MBEDTLS_PSA_KEY_STORE_DYNAMIC` is enabled, destroying a key might not be perfectly symmetric with creating a key, because the key table grows dynamically and may become fragmented and has some hysteresis.)

In Mbed TLS tests, we ensure that each test case frees all the memory that it allocates, and we validate each test suite with Asan. Tests that use the PSA crypto API start with `psa_crypto_init()` and end with `mbedtls_psa_crypto_free()`. The call to `mbedtls_psa_crypto_free()` destroys any existing key, so to ensure that higher-level APIs don't leak key objects, [we fail the test if the key store is not empty before calling `mbedtls_psa_crypto_free()`](https://github.com/Mbed-TLS/mbedtls-framework/blob/1a83e0c84d4b7aa11c7cfd3771322486fc87d281/tests/include/test/psa_crypto_helpers.h#L26) by calling [`mbedtls_test_fail_is_psa_leaking()`](https://github.com/Mbed-TLS/mbedtls-framework/blob/1a83e0c84d4b7aa11c7cfd3771322486fc87d281/tests/src/psa_crypto_helpers.c#L71).

One more consideration is that when CTR_DRBG is enabled and AES is fully accelerated (`MBEDTLS_AES_C` disabled), the key used by the DRBG is still present in the key store. The code of `mbedtls_test_helper_is_psa_leaking()` accounts for that. I'm not sure if that's supported in OpenThread though.

LuDuda replied:
Introduced a new APIs `otPlatCryptoCAlloc` and `otPlatCryptoFree` as discussed with Abtin. That allows me to remove this workarounds in the tests.

LuDuda commented on src/core/meshcop/tcat_agent.cpp:
```diff
@@ -903,12 +903,31 @@ void TcatAgent::CalculateHash(uint64_t aChallenge, const char *aBuf, size_t aBuf
+    Crypto::Storage::KeyRef keyRef;
+    Error                   error;
+
+    error = Crypto::Storage::ImportKey(keyRef, Crypto::Storage::kKeyTypeHmac, Crypto::Storage::kKeyAlgorithmHmacSha256,
```
Not really, I can combine these lines into single line.

LuDuda committed:
[crypto] PSA API: remove otPlatCryptoInit API

This commit removes otPlatCryptoInit API and moves responsibility of
initializing the Crypto subsystem to the platform.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

testter2 commented:
Thanks for this update!  I’m in the process of updating my platform port of OpenThread and would like to prepare in advance for when this PR gets merged.

Could you clarify what changes will be expected from platform implementations after this PR?
Specifically:

Will it be mandatory for platforms to provide their own otPlatCryptoImportKey / otPlatCryptoExportKey along with a secure storage backend as defaults?

Are there any mandatory configuration flags or migration steps we should plan for when enabling this?

This will help us understand how to adapt our platform layer to support PSA ITS correctly once the change lands. Thanks!

LuDuda committed:
[crypto] PSA API: support HKDF extraction with HMAC-SHA256 keys

The PSA platform now transparently handles HKDF operations when using
HMAC-SHA256 keys by exporting and re-importing them as volatile
HKDF-SHA256 keys. The export/import logic has been moved into the
platform layer. This change is required to support TREL and the PSA API
key usage restrictions.

Signed-off-by: Łukasz Duda <lukasz.duda@nordicsemi.no>

gilles-peskine-arm commented on third_party/mbedtls/native_its/src/file/its_file.c:
```diff
@@ -0,0 +1,384 @@
+
+/**
+ * @file
+ *   This file implements an exemplary ITS working on files.
```
The whole `its_file.c` seems very similar to Mbed TLS's `psa_its_file.c`, except with support for a directory that's configurable at runtime. That's a feature that we'd like to have but never got around to implementing. Could we keep our efforts in a single implementation? Or do you need to have your own to avoid having Mbed TLS or TF-PSA-Crypto as a dependency?

gilles-peskine-arm commented on third_party/mbedtls/native_its/src/file/its_file.c:
```diff
@@ -0,0 +1,384 @@
+exit:
+    if (file)
+    {
+        fclose(file);
```
If `fclose` fails, in standard C, it's possible that the file hasn't been fully written. I don't know if that's a concern on OpenThread.

Damian-Nordic reviewed (commented):
Comments to the newest commit.

LuDuda commented on src/core/crypto/crypto_platform_psa.cpp:
```diff
@@ -463,10 +469,36 @@ OT_TOOL_WEAK otError otPlatCryptoHkdfExtract(otCryptoContext   *aContext,
+    }
+
 exit:
+    if (keyRef != PSA_KEY_ID_NULL)
```
Thanks, fixed.

LuDuda commented on src/core/crypto/crypto_platform_psa.cpp:
```diff
@@ -451,9 +451,15 @@ OT_TOOL_WEAK otError otPlatCryptoHkdfExtract(otCryptoContext   *aContext,
+    psa_key_attributes_t            attributes  = PSA_KEY_ATTRIBUTES_INIT;
+    psa_algorithm_t                 keyAlg      = PSA_ALG_NONE;
+    size_t                          keyLength   = 0;
+    constexpr size_t                kBufferSize = 128;
```
This is a valid comment. However, while it is not ideal, I still want to keep the unit tests that use test vectors from RFC 5869, so I have left 80 bytes here.

abtink commented on tests/unit/test_dns_client.cpp:
```diff
@@ -135,6 +135,9 @@ void otPlatFree(void *aPtr)
 }
+
+void *otPlatCryptoCAlloc(size_t aNum, size_t aSize) { return calloc(aNum, aSize); }
+void  otPlatCryptoFree(void *aPtr) { free(aPtr); }
```
Can we define these in the `tests/unit/test_platform.cpp`?
This way it will be available for all unit tests to use?

abtink commented on tests/unit/test_hmac_sha256.cpp:
```diff
@@ -239,12 +244,23 @@ void TestHmacSha256(void)
         Crypto::HmacSha256::Hash hash;
+        Crypto::Key              key;
+
+#if OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
```
Some suggestions here:

* For each `testCase`, we import the key into PSA and obtain a temporary key-ref, which we save in `testCase.mKeyRef` (I assume this is why you removed the `const` from the `for` loop above).
* Do we really need to save this key-ref in the `testCase` struct?
* Can we just save it in a local variable here within the loop block?
* At the end of the loop, we can call `Crypto::Storage::DestroyKey(keyref)` to remove the temporary key-ref.

We need to do the same in the other `for()` loop iterating over `kTestCases`, but I think that is fine (and even more clear). 
- Each block executes its own logic and is stand-alone (importing its own key into storage and obtaining its own temporary key-ref, destrory temp key-ref at end of the loop).
-  Attempting to save the `mKeyRef` in the `testCase` struct to share or reuse the same key-ref between `for` loops makes it more complex to understand the code.

Thoughts?

abtink commented on tests/unit/test_platform.cpp:
```diff
@@ -226,6 +230,36 @@ OT_TOOL_WEAK otError otPlatEntropyGet(uint8_t *aOutput, uint16_t aOutputLength)
 }
 
+#if (OPENTHREAD_CONFIG_CRYPTO_LIB == OPENTHREAD_CONFIG_CRYPTO_LIB_PSA) && defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
+/**
```
I don't think you want this comment/description in the `doxygen` documentation (e.g., on the `openthread.io` website), so I suggest not using the `/**` format. Perhaps we can move it inside the function itself?

abtink commented on .github/workflows/toranj.yml:
```diff
@@ -180,9 +180,6 @@ jobs:
         ./tests/toranj/build.sh --log-level NONE all
-        #- - - - - - - - - - - - - - - - - - - - - - - - - - -
-        git clean -dfx
-        ./tests/toranj/build.sh --enable-plat-key-ref all
```
I see you removed this.

I added this back in 
- #9620.

From what I recall, the issue was that after the KEY_REF code was contributed to OT, there was no GitHub CI test covering this, or at a minimum, checking whether the OT stack would compile correctly (with no errors) when this feature is enabled. This is important because as new PRs are merged, we want to ensure we at least build with this feature enabled so that no new PR breaks the build.

If you are removing this, I assume we are now going to cover this feature in one of the newly added CI jobs? Just asking here to confirm. Thanks. 😃

abtink commented on src/core/crypto/crypto_platform_psa.cpp:
```diff
@@ -0,0 +1,795 @@
+    case PSA_ERROR_BUFFER_TOO_SMALL:
+        return kErrorNoBufs;
+    default:
+        return kErrorFailed;
```
Should follow single `return` from any function/method (per our style guide)?
There are multiple instances of this  this in this source file.

abtink commented on include/openthread/platform/crypto.h:
```diff
@@ -212,11 +215,6 @@ typedef struct otPlatCryptoEcdsaSignature otPlatCryptoEcdsaSignature;
-/**
- * Initialize the Crypto module.
- */
-void otPlatCryptoInit(void);
```
I noticed you removed this existing `otPlat` API. Is there a reason for this?

We should keep existing APIs unchanged to ensure backward compatibility. Some platforms may perform operations in this API and expect it to be called by the OT stack. Removing it may go unnoticed and cause issues.

I suggest keeping it, even if it ends up being an empty function in most platform implementations.

-----

We should be very careful when designing or adding new public or platform APIs. Adding is always easy, but changing or removing is almost impossible.

LuDuda replied:
This has been discussed on the Chat and here https://github.com/openthread/openthread/pull/11445#discussion_r2484785986

The context was adding support for RCP (which is not part of this PR) - which needs PSA API initialization, but not from KeyManager. I can leave it as of now, and we can later decide how OT could call `otPlatCryptoInit()` if we want to preserve this.

abtink reviewed (commented):
Thanks, @LuDuda , for all this work.

I have reviewed a good chunk of this PR (not everything) and added some suggestions below.

It is a big PR with a large `git diff`. It may be helpful if we can break it into smaller PRs, each focusing on a smaller change. This would make it easier to review and move forward (merging smaller pieces first) more quickly.

Here are some ideas for parts that could be broken into their own smaller PRs:

* The introduction of `otPlatCryptoCAlloc()` and `otPlatCryptoFree` (optional platform APIs) could be its own smaller PR.
* The changes to the unit tests to make them work under `KEY_REF` may be a good thing to separate out.
   * Maybe we can even add an implementation of the related `otPlatCrypto` APIs (like `otPlatCryptoImportKey`, `otPlatCryptoExportKey`, `otPlatCryptoDestroyKey()`, etc.) under the simulation platform to emulate these (saving them in an array?).
   * This way, we can run unit tests under simulation without requiring the full PSA implementation (which is what we do for radio, alarm, settings, and other modules).

* Another part that can be added as its own is the introduction of `OT_CRYPTO_KEY_TYPE_DERIVE` and its related code.
* Then, finally, we can have a PR that handles `OPENTHREAD_CONFIG_CRYPTO_LIB` and the PSA related platform code.

What do you think?

----

Edit: 

@LuDuda, I noticed you have already staged this PR beautifully as a series of separate smaller commits (each making a smaller change). Could we perhaps break somer of these into smaller PRs? 

(mentioned the same here ->https://github.com/openthread/openthread/pull/11445#issuecomment-3657859525)

zwx1995esp commented on src/core/crypto/mbedtls.cpp:
```diff
@@ -57,9 +57,9 @@ MbedTls::MbedTls(void)
+#endif // OPENTHREAD_CONFIG_ENABLE_BUILTIN_MBEDTLS_MANAGEMENT && !OPENTHREAD_CONFIG_HEAP_EXTERNAL_ENABLE
 }
 
 Error MbedTls::MapError(int aMbedTlsError)
```
And for these two macro, they used same value `PSA_ERROR_INVALID_ARGUMENT`

[MBEDTLS_ERR_SSL_BAD_INPUT_DATA](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/ssl.h#L40)
[MBEDTLS_ERR_PK_BAD_INPUT_DATA](https://github.com/Mbed-TLS/TF-PSA-Crypto/blob/1e81ba41a12e75f26fe66877db06249bfbaef59e/include/mbedtls/compat-3-crypto.h#L37)



These three used same value `PSA_ERROR_INSUFFICIENT_MEMORY`
[MBEDTLS_ERR_PEM_ALLOC_FAILED](https://github.com/Mbed-TLS/TF-PSA-Crypto/blob/1e81ba41a12e75f26fe66877db06249bfbaef59e/include/mbedtls/compat-3-crypto.h#L30)
[MBEDTLS_ERR_X509_ALLOC_FAILED](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/x509.h#L73)
[MBEDTLS_ERR_SSL_ALLOC_FAILED](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/ssl.h#L101)

The same value cannot be used in the `switch` logic. Could we separate this file to handle PSA and non-PSA scenarios? In the PSA scenario, PSA_ERROR_XXX can be mapped to OpenThread error codes.

zwx1995esp commented:
Also for the function in the file src/core/meshcop/secure_transport.cpp: [mbedtls_ssl_conf_rng](https://github.com/openthread/openthread/blob/main/src/core/meshcop/secure_transport.cpp#L202) this API is removed in mbedtls 4.0.

and for   [the line 281 in the file secure_transport.cpp](https://github.com/openthread/openthread/blob/main/src/core/meshcop/secure_transport.cpp#L281), [mbedtls_ssl_cookie_setup](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/ssl_cookie.h#L58) now has only one parameter in mbedtls 4.0.

We should also consider these two APIs breaking changes.

LuDuda commented on src/core/config/platform.h:
```diff
@@ -154,9 +154,13 @@
 #ifndef OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
-#define OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE 0
+#define OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE \
+    (OPENTHREAD_CONFIG_CRYPTO_LIB == OPENTHREAD_CONFIG_CRYPTO_LIB_PSA)
```
PSA Crypto requires key references and does not operate on raw key material. What is your use case for supporting PSA Crypto without OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE? Would you like to create ad-hoc volatile keys on the platform layer each time a cryptographic operation is performed? This would defeat one of the main advantages of the PSA Crypto design.

As requested, I will start contributing this PR in parts this week.

LuDuda replied:
Yes, RCP support is not part of this PR. I had it in my internal branch along with POSIX support (as far as i remember, it was mostly about heap support), but I wanted to start with the simulation platform first. Nevertheless, PSA support for RCP is already available in some silicon vendors’ SDKs.

LuDuda commented:
@hemanth-silabs Yes, this topic has been discussed above, and three options have been outlined (https://github.com/openthread/openthread/pull/11445#discussion_r2484786271).

As suggested by Abtin, I've implemented option 2 which is included in the last commit of this PR.
